File: sdk/javascript/src/cache.ts
import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';
import { homedir } from 'os';

interface CacheEntry {
  text_hash: string;
  source_lang: string;
  target_lang: string;
  original_text: string;
  translated_text: string;
  provider: string;
  created_at: number;
}

interface CacheData {
  entries: Record<string, CacheEntry>;
}

export class JsonCache {
  private cachePath: string;
  private data: CacheData;
  private maxAge: number;

  constructor(cacheDir?: string, maxAge: number = 86400000) {
    const dir = cacheDir || path.join(homedir(), '.source-translator');
    this.cachePath = path.join(dir, 'cache.json');
    this.maxAge = maxAge;
    this.data = this.load();
  }

  private load(): CacheData {
    try {
      if (fs.existsSync(this.cachePath)) {
        const content = fs.readFileSync(this.cachePath, 'utf-8');
        return JSON.parse(content);
      }
    } catch {
      // Ignore errors
    }
    return { entries: {} };
  }

  private save(): void {
    try {
      const dir = path.dirname(this.cachePath);
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
      fs.writeFileSync(this.cachePath, JSON.stringify(this.data, null, 2));
    } catch (err) {
      console.error('Failed to save cache:', err);
    }
  }

  private computeHash(text: string, targetLang: string): string {
    return crypto.createHash('sha256')
      .update(text)
      .update(targetLang)
      .digest('hex');
  }

  get(text: string, targetLang: string): CacheEntry | null {
    const hash = this.computeHash(text, targetLang);
    const entry = this.data.entries[hash];
    
    if (!entry) return null;
    
    if (Date.now() - entry.created_at > this.maxAge) {
      delete this.data.entries[hash];
      this.save();
      return null;
    }
    
    return entry;
  }

  set(text: string, sourceLang: string, targetLang: string, translation: string, provider: string): void {
    const hash = this.computeHash(text, targetLang);
    this.data.entries[hash] = {
      text_hash: hash,
      source_lang: sourceLang,
      target_lang: targetLang,
      original_text: text,
      translated_text: translation,
      provider,
      created_at: Date.now(),
    };
    this.save();
  }

  stats(): { total_entries: number; providers_used: string[] } {
    const providers = new Set<string>();
    for (const entry of Object.values(this.data.entries)) {
      providers.add(entry.provider);
    }
    return {
      total_entries: Object.keys(this.data.entries).length,
      providers_used: Array.from(providers),
    };
  }
}

File: sdk/python/source_translator/cache.py
import hashlib
import json
import time
from pathlib import Path
from typing import Optional
from dataclasses import dataclass, asdict


@dataclass
class CacheEntry:
    text_hash: str
    source_lang: str
    target_lang: str
    original_text: str
    translated_text: str
    provider: str
    created_at: float


class JsonCache:
    def __init__(self, cache_dir: Optional[str] = None, max_age: int = 86400):
        if cache_dir:
            self.cache_path = Path(cache_dir) / "cache.json"
        else:
            self.cache_path = Path.home() / ".source-translator" / "cache.json"
        
        self.max_age = max_age
        self.data: dict[str, CacheEntry] = {}
        self._load()

    def _load(self) -> None:
        try:
            if self.cache_path.exists():
                with open(self.cache_path, "r") as f:
                    raw = json.load(f)
                    for key, entry in raw.items():
                        self.data[key] = CacheEntry(**entry)
        except Exception:
            pass

    def _save(self) -> None:
        try:
            self.cache_path.parent.mkdir(parents=True, exist_ok=True)
            with open(self.cache_path, "w") as f:
                json.dump(
                    {k: asdict(v) for k, v in self.data.items()},
                    f,
                    indent=2,
                )
        except Exception:
            pass

    def _compute_hash(self, text: str, target_lang: str) -> str:
        return hashlib.sha256(f"{text}{target_lang}".encode()).hexdigest()

    def get(self, text: str, target_lang: str) -> Optional[CacheEntry]:
        hash_key = self._compute_hash(text, target_lang)
        entry = self.data.get(hash_key)
        
        if entry is None:
            return None
        
        if time.time() - entry.created_at > self.max_age:
            del self.data[hash_key]
            self._save()
            return None
        
        return entry

    def set(self, text: str, source_lang: str, target_lang: str, translation: str, provider: str) -> None:
        hash_key = self._compute_hash(text, target_lang)
        self.data[hash_key] = CacheEntry(
            text_hash=hash_key,
            source_lang=source_lang,
            target_lang=target_lang,
            original_text=text,
            translated_text=translation,
            provider=provider,
            created_at=time.time(),
        )
        self._save()

    def stats(self) -> dict:
        providers = set()
        for entry in self.data.values():
            providers.add(entry.provider)
        return {
            "total_entries": len(self.data),
            "providers_used": list(providers),
        }
